src/plugins/intel_cpu/src/proxy_mem_mgr.cpp

```diff
@@ -0,0 +1,66 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "proxy_mem_mgr.h"
+#include "utils/debug_capabilities.h"
+
+using namespace ov::intel_cpu;
+
+void ProxyMemoryMngr::reset(std::shared_ptr<IMemoryMngr> _pMngr) {
+    auto _validated = (_pMngr != m_pMngr);
+    if (_pMngr) {
+        m_pMngr = _pMngr;
+    } else {
+        m_pMngr = m_pOrigMngr;
+    }
+
+    // WA: unconditionally resize to last size
+    if (_validated) {
+        auto res = m_pMngr->resize(m_size);
+        DEBUG_LOG(this, ", ", m_pMngr, " size ", m_size, " -> ", m_size, " resized? ", res, " RawPtr ", getRawPtr());
+
+        notifyUpdate();
+    }
+}
```

Comment:
1. The ask was to split the semantics
2. Notify update must be called on **ANY** modification!
```cpp
void ProxyMemoryMngr::setMemMngr(std::shared_ptr<IMemoryMngr> pMngr) {
    OPENVINO_ASSERT(pMngr, "Attempt to set empty manager");
    if (m_pMngr == pMngr) {
        return;
    }
    m_pMngr = pMngr;
    m_pMngr->resize(m_size);
    notifyUpdate();
}

void reset() {
    if (m_pMngr == m_pOrigMngr) {
        return;
    }
    m_pMngr = m_pOrigMngr;
    m_pMngr->resize(m_size);
    notifyUpdate();
}
```

Reply:
What's your idea of m_pOrigMngr? Is it the original actual manager to create the graph memory object? In the target problem, I think it is. But if we create it with the explicit constructor, it won't be, because m_pOrigMngr is instanced by the constructor without parameters, and m_pMngr is another instance.

                        std::make_shared<ProxyMemoryMngr>(std::make_shared<DnnlMemoryMngr>(make_unique<MemoryMngrWithReuse>()));

                        std::make_shared<ProxyMemoryMngr>();

It wouldn't arouse a big problem except an extra memory manager instanced.

Reply:
The coment above has nothing to do with the constructor, this is simply a refactoring of the original `void ProxyMemoryMngr::reset(std::shared_ptr<IMemoryMngr> _pMngr)` method.
1. We split responsibility: `ProxyMemoryMngr::setMemMngr` does only the part of setting the external mem mngr, and `reset()` bring back the default one instead of calling `reset(nullptr_t)` in your version
2. We call `notifyUpdate` on every change of the actual underlining memory manager, actually this part is true for you solution as well, but the condition `_validated ` is not sat robust for future changes.